src/shims/x86/gfni.rs

```diff
@@ -0,0 +1,198 @@
+        }
+
+        // Multiply the matrix with the vector and perform the addition.
+        for j in 0..8 {
+            let index = i.wrapping_add(j);
+            let left = this.read_scalar(&this.project_index(&left, index)?)?.to_u8()?;
+            let left = if inverse { TABLE[usize::from(left)] } else { left };
+
+            let mut res = 0;
+
+            // Do the matrix multiplication.
+            for bit in 0u8..8 {
+                let mut b = matrix[usize::from(bit)] & left;
+
+                // Calculate the parity bit.
+                b = (b & 0b1111) ^ (b >> 4);
+                b = (b & 0b11) ^ (b >> 2);
+                b = (b & 0b1) ^ (b >> 1);
+
+                res |= b << 7u8.wrapping_sub(bit);
+            }
+
+            // Perform the addition.
+            res ^= imm8;
+
+            let dest = this.project_index(&dest, index)?;
+            this.write_scalar(Scalar::from_u8(res), &dest)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// A lookup table for computing the inverse byte for the inverse affine transformation.
+// This is a evaluated at compile time. Trait based conversion is not available.
+/// See <https://www.corsix.org/content/galois-field-instructions-2021-cpus> for the
+/// definition of `gf_inv` which was used for the creation of this table.
+#[allow(clippy::cast_possible_truncation)]
+static TABLE: [u8; 256] = {
+    let mut array = [0; 256];
+
+    let mut i = 1;
+    while i < 256 {
+        let mut x = i as u8;
+        let mut y = gf2p8_mul(x, x);
+        x = y;
+        let mut j = 2;
+        while j < 8 {
+            x = gf2p8_mul(x, x);
+            y = gf2p8_mul(x, y);
+            j += 1;
+        }
+        array[i] = y;
+        i += 1;
+    }
+
+    array
+};
+
+/// Multiplies packed 8-bit integers in `left` and `right` in the finite field GF(2^8)
+/// and store the results in `dst`. The field GF(2^8) is represented in
+/// polynomial representation with the reduction polynomial x^8 + x^4 + x^3 + x + 1.
+/// See <https://www.corsix.org/content/galois-field-instructions-2021-cpus> for details.
+// This is a const function. Trait based conversion is not available.
+#[allow(clippy::cast_possible_truncation)]
+const fn gf2p8_mul(left: u8, right: u8) -> u8 {
+    // The reduction polynomial used for the multiplication.
+    const POLYNOMIAL: u32 = 0x11b;
+
+    let left = left as u32;
+    let right = right as u32;
+
+    let mut result = 0u32;
+
+    // Step 1: Carryless multiplication of `left` and `right`.
+    // Carryless multiplication is explained in the `x86::pclmulqdq` function.
+    let mut i = 0u32;
+    while i < 8 {
+        if left & (1 << i) != 0 {
+            result ^= right << i;
+        }
+        i = i.wrapping_add(1);
+    }
+
+    // Step 2: Polynomial division modulo 2.
+    // Polynomial division modulo 2 is defined inside the `x86::sse42` module
+    // inside the "_mm_crc32_u{8, 16, 32, 64}" intrinsics implementation.
```

Comment:
This looks quite a bit different from that implementation though. I can only pattern-match here since I don't understand the code, but other than "it's a loop with XOR and shift", it seems unrelated to what `_mm_crc32_u*` does.

If this is the same procedure, can it be extracted into a helper function?

Reply:
Algorithmically, both the crc32 intrinsics and the finite field multiplication use polynomial division modulo 2 in their definition, so the loops do the same thing. Maybe "described" is a better word than "defined".

The difference between the implementations stems from the fact that the intrinsics guide presents this step differently between the crc32 and gfni intrinsics. In case of the the crc32 intrinsics, the step is reduced to an unspecified `MOD2` function call, while the gfni intrinsics provides a full loop.

A nice property of this implementation is the one-to-one correspondence between the implementation and the definition of the intrinsics guide: Comparing the intrinsics guide definition with this one will make the correctness aspect very obvious. While this step can be extracted into a separate function, that correspondence would be lost.

If it should be extracted, where should the function be placed?

Reply:
> A nice property of this implementation is the one-to-one correspondence between the implementation and the definition of the intrinsics guide

Where can I find that definition?

Reply:
One of the entries in https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=gf2p8mul.
I can add the link to the implementation.

Reply:
Yes, that would probably be a better link than `_mm_crc32_u*` then.

Reply:
Done.